crates/bevy_camera/src/frame_capture.rs

```diff
@@ -0,0 +1,363 @@
+use bevy_app::prelude::*;
+use bevy_asset::{Assets, Handle};
+use bevy_core_pipeline::{draw_3d_graph, node, AlphaMask3d, Opaque3d, Transparent3d};
+use bevy_ecs::entity::Entity;
+use bevy_ecs::prelude::{Component, World};
+use bevy_ecs::system::{Commands, Query};
+
+use bevy_render::prelude::Image;
+use bevy_render::render_asset::RenderAssets;
+use bevy_render::render_graph::{self, NodeRunError, RenderGraph, RenderGraphContext, SlotValue};
+use bevy_render::render_phase::RenderPhase;
+use bevy_render::render_resource::{
+    Buffer, BufferDescriptor, BufferUsages, CommandEncoderDescriptor, Extent3d, ImageCopyBuffer,
+    ImageDataLayout, TextureDescriptor, TextureDimension, TextureFormat, TextureUsages,
+};
+use bevy_render::renderer::{RenderContext, RenderDevice, RenderQueue};
+use bevy_render::{RenderApp, RenderStage};
+use wgpu::{BufferView, MapMode};
+
+// The name of the final node of the first pass.
+pub const FRAME_CAPTURE_DRIVER: &str = "frame_capture_driver";
+
+#[derive(Clone)]
+pub struct CPUBuffer(pub Buffer);
+
+impl CPUBuffer {
+    pub fn get<F>(&self, render_device: &RenderDevice, f: F)
+    where
+        F: FnOnce(BufferView),
+    {
+        let large_buffer_slice = self.0.slice(..);
+        render_device.map_buffer(&large_buffer_slice, MapMode::Read);
+        {
+            let large_padded_buffer = large_buffer_slice.get_mapped_range();
+
+            f(large_padded_buffer);
+        }
+        self.0.unmap();
+    }
+}
+
+#[derive(Clone)]
+pub enum TargetBuffer {
+    CPUBuffer(CPUBuffer),
+    GPUBuffer(Handle<Image>),
+}
+
+#[derive(Component, Clone)]
+pub struct FrameCapture {
+    pub target_buffer: Option<TargetBuffer>,
+    pub gpu_image: Handle<Image>,
+    pub width: u32,
+    pub height: u32,
+    pub camera: Option<Entity>,
+    pub active: bool,
+}
+
+impl FrameCapture {
+    pub fn new_cpu_buffer(
+        width: u32,
+        height: u32,
+        active: bool,
+        format: TextureFormat,
+        images: &mut Assets<Image>,
+        render_device: &RenderDevice,
+    ) -> Self {
+        let size = Extent3d {
+            width,
+            height,
+            ..Default::default()
+        };
+
+        // This is the texture that will be rendered to.
+        let mut image = Image {
+            texture_descriptor: TextureDescriptor {
+                label: None,
+                size,
+                dimension: TextureDimension::D2,
+                format,
+                mip_level_count: 1,
+                sample_count: 1,
+                usage: TextureUsages::TEXTURE_BINDING
+                    | TextureUsages::COPY_DST
+                    | TextureUsages::COPY_SRC
+                    | TextureUsages::RENDER_ATTACHMENT,
+            },
+            ..Default::default()
+        };
+        image.resize(size);
+
+        let gpu_image = images.add(image);
+
+        let padded_bytes_per_row = RenderDevice::align_copy_bytes_per_row(width as usize) * 4;
```

Comment:
I'm not happy about the `* 4` there (and the other time there's a 4). Some texture formats have less than 4 components, and some uses more than one byte per component.
Should this be `* format.describe().components as usize * format.describe().block_size as usize` instead?

Reply:
Good catch.

Based on this: [ImageDataLayout](https://wgpu.rs/doc/wgpu/struct.ImageDataLayout.html) I think it may be:

```rs
let padded_bytes_per_row = RenderDevice::align_copy_bytes_per_row(
    (gpu_image.size.width as usize / format.block_dimensions.0 as usize) * format.block_size as usize,
);
```

[This](https://github.com/bevyengine/bevy/blob/00f83941b16e7a175c5c1503be81a648e23d81df/crates/bevy_render/src/texture/image.rs#L582) is what Bevy does for `prepare_asset` in `image.rs`

And it looks like `pixel_size()` just does `info.type_size * info.num_components`. I'm not sure if this ends up being always correct if the texture is compressed.

There's also the issue of the buffer being larger than the user might expect since it's [aligned to 256 bytes](https://docs.rs/wgpu-types/latest/wgpu_types/constant.COPY_BYTES_PER_ROW_ALIGNMENT.html)

This is fine for both GPU examples since it's what the gpu expects on both ends. But the one that copies to the CPU results in row padded data.

Do we crop the buffer for the user? If so, I'm not sure exactly what the most efficient way of doing that is. Each row is potentially too long, so we can't just slice off the end of the flat buffer.